File: Core-backend-layer/Programming-fundamentals/Python/Threads/src/product.py
import threading

class Product:
    def __init__(self, name, quantity):
        self.name = name
        self.quantity = quantity

    def get_name(self):
        return self.name

    def get_quantity(self):
        return self.quantity

    def set_quantity(self, quantity):
        self.quantity = quantity

    def decrease_quantity(self, amount):
        # Using Lock to avoid potential data inconsistencies when multiple threads access shared resources concurrently
        with threading.Lock():
            if 0 < amount <= self.quantity:
                self.quantity -= amount
                print(f"Ordered {amount} units of {self.name}. Remaining quantity: {self.quantity}")
            else:
                print(f"Invalid order quantity for {self.name}")

File: Core-backend-layer/Programming-fundamentals/Java/Threads/src/Product.java
public class Product {
    private String name;
    private int quantity;

    public Product(String name, int quantity) {
        this.name = name;
        this.quantity = quantity;
    }

    public String getName() {
        return name;
    }

    public int getQuantity() {
        return quantity;
    }

    public void setQuantity() {
        this.quantity = quantity;
    }

    public void decreaseQuantity(int amount) {
        // synchronized helps avoid potential data inconsistencies when multiple threads access shared resources concurrently
        synchronized (this) {
            if (amount > 0 && amount <= quantity) {
                quantity -= amount;
                System.out.println("Ordered " + amount + " units of " + name + ". Remaining quantity: " + quantity);
            } else {
                System.out.println("Invalid order quantity for " + name);
            }
        }
    }
}
